# NumPy
<li>NumPy stands for <b>Numerical Python</b>.</li>
<li>NumPy is the fundamental packages for scientific computing in Python.</li>
<li>It is a Python library which provides support for large, multi-dimensional arrays and matrices.</li>
<li>It also provides support for a large collection of high-level mathematical functions to operate on these arrays and matrices.</li>


## Importance Of NumPy In Python

<li>support for wide variety of mathematical operation on arrays and matrices.</li>
<li>provides various functions for searching, sorting, performing statistical operations and other basic linear algebric operations.</li>

## Installation Of NumPy
<li>Go to your terminal, open and activate your virtual environment and then use the following commands for installing numpy.</li>
<code>
    pip install numpy
</code>


## Importing NumPy
<li>We need to import numpy if we want to create any numpy arrays and perform any operations on them.</li>
<li>We can import numpy package using the following command:</li>
<code>
import numpy as np
</code>

In [1]:
import numpy as np

## Creating Numpy Arrays.

<li>We can create a numpy array by using the function np.array()</li>
<li>Inside the np.array() method, we pass a list of items or a list of list of items.</li>

In [2]:
numpy_array = np.array([1,2,3,4])

In [3]:
print(numpy_array)

[1 2 3 4]


In [5]:
print(type(numpy_array))

<class 'numpy.ndarray'>


In [4]:
a = [1,2,3,4]
print(a)

[1, 2, 3, 4]


In [6]:
print(type(a))

<class 'list'>


In [7]:
list_of_list = [['Ram', 'Thapa', 22], ['Shyam', 'Thapa', 32]]

In [8]:
twod_array = np.array(list_of_list)
print(twod_array)

[['Ram' 'Thapa' '22']
 ['Shyam' 'Thapa' '32']]


In [9]:
print(type(twod_array))

<class 'numpy.ndarray'>


### Question

<li>Import numpy, and assign it to the alias np.</li>
<li>Create a NumPy ndarray from the list [10, 20, 30]. Assign the result to the variable data_ndarray</li>

In [10]:
import numpy as np

In [11]:
data_ndarray = np.array([10,20,30])

In [12]:
print(data_ndarray)

[10 20 30]


In [13]:
print(type(data_ndarray))

<class 'numpy.ndarray'>


## NumPy Vs Python Lists
<li>Importing module</li>
<li>Numerical Operations.</li>
<li>consumes less memory.</li>
<li>fast as compared to the python itselt.</li>
<li>convenient to use.</li>

## Comparing Numpy With Python Lists

<li><b>Size/Memory:</b>Numpy array takes less space than that of Python lists.</li>
<li><b>Speed/Performance:</b>It takes less time in Numpy array to compute the same task than that of Python lists.</li>

### Memory Comparision Between NumPy & Python Lists
<li>For doing the memory comparision, we have getsizeof() function which can be imported from sys library.</li>
<li>We can create a list and numpy array containing the same elements and check their size and compare them.</li>


In [14]:
from sys import getsizeof

In [25]:
x_list = [i for i in range(1000000)]
size_xlist = getsizeof(x_list)
print(size_xlist)

8697456


In [26]:
x_array = np.array(x_list)
size_xarray = getsizeof(x_array)
print(size_xarray)

4000112


In [29]:
print("NumPy array takes {} times less space than that of Python list".format(size_xlist / size_xarray))

NumPy array takes 2.1743031195126536 times less space than that of Python list


### Speed Comparision Between NumPy Array & Python List
<li>For doing the speed comparision, we have to perform same operation with Python list and NumPy arrays.</li>
<li>We can calculate the time taken by each operation and then compare their speed or performance.</li>
<li>For doing this, we have a time module(library) from which we can calculate the start_time using time.process_time().</li>
<li>Similarly, we can also calculate the end_time using the same time.process_time().</li>
<li>Here start_time means time before performing operation and end_time mens time after performing operation.</li>
<li>This way, we can calculate the time taken by the operation by using end_time - start_time formula.</li>
<li>Finally, we can compare both times and see which is faster.</li>

In [30]:
first_list = [i for i in range(1, 100000000)]

In [31]:
print(type(first_list))

<class 'list'>


In [32]:
first_ndarray = np.array(first_list)

In [33]:
print(type(first_ndarray))

<class 'numpy.ndarray'>


In [34]:
import time

In [35]:
start_time = time.perf_counter()
sum_first_list = sum(first_list)
end_time = time.perf_counter()
list_processing_time = end_time - start_time
print(list_processing_time)

1.7313156000000163


In [36]:
start_time = time.perf_counter()
sum_first_array = np.sum(first_ndarray)
end_time = time.perf_counter()
ndarray_processing_time = end_time - start_time
print(ndarray_processing_time)

0.05772589999992306


In [37]:
print("Operations On NumPy is {} time faster than that of Python list".format(list_processing_time/ndarray_processing_time))

Operations On NumPy is 29.992007054066267 time faster than that of Python list


## Dimensions In Arrays
<li>You can get the number of dimensions, shape (length of each dimension), and size (number of all elements) of the NumPy array with ndim, shape, and size attributes of numpy.ndarray.</li>
<li>The built-in function len() returns the size of the first dimension.</li>
<li>Number of dimensions of the NumPy array: ndim</li>
<li>Shape of the NumPy array: shape</li>
<li>Size of the NumPy array: size</li>
<li>Size of the first dimension of the NumPy array: len()</li>

<ol>
    <li><b>One Dimensional Arrays:</b> If an array has only one dimension then it is one dimensional array.</li>
    <li><b>Two Dimensional Arrays:</b> If an array has two dimensions along x and y axis then it is two dimensional arrays.</li>
    <li><b>Three Dimensional Arrays:</b> If an array has three dimensions along x, y and z axis then it is three dimensional arrays.</li>
    <li><b>Higher(N) Dimensional Arrays:</b> If an array has more than three dimensions then such arrays are called n-dimensional  or higher dimensional arrays.</li>
</ol>

In [38]:
onedarray = np.array([1,2,3,4])

In [39]:
onedarray.ndim

1

In [45]:
onedarray.shape

(4,)

In [47]:
onedarray.size

4

In [46]:
len(onedarray)

4

In [48]:
twodarray = np.array([[1,2,3,4], [5,6,7,8]])

In [49]:
print(twodarray)

[[1 2 3 4]
 [5 6 7 8]]


In [50]:
twodarray.ndim

2

In [51]:
twodarray.size

8

In [52]:
twodarray.shape

(2, 4)

In [53]:
len(twodarray)

2

In [54]:
threedarray = np.array([[[1, 2, 3], [4, 5, 6]]])

In [55]:
threedarray.ndim

3

In [56]:
threedarray.shape

(1, 2, 3)

In [57]:
threedarray.size

6

In [58]:
len(threedarray)

1

In [59]:
threedarray2 = np.array([[[1,2,3], [4,5,6]], [[9,8,7], [6,5,4]]])

In [62]:
print(threedarray2)

[[[1 2 3]
  [4 5 6]]

 [[9 8 7]
  [6 5 4]]]


In [60]:
threedarray2.ndim

3

In [61]:
threedarray2.shape

(2, 2, 3)

In [63]:
threedarray2.size

12

In [64]:
len(threedarray2)

2

In [67]:
fourdarray = np.array([[[[1,2,3,4], [4,3,2,1]]], [[[5,6,7,8], [8,7,6,5]]]])

In [71]:
print(fourdarray)

[[[[1 2 3 4]
   [4 3 2 1]]]


 [[[5 6 7 8]
   [8 7 6 5]]]]


In [72]:
fourdarray.ndim

4

In [73]:
fourdarray.shape

(2, 1, 2, 4)

In [74]:
fourdarray.size

16

In [75]:
len(fourdarray)

2

In [76]:
fivedarray = np.array([1,2,3,4], ndmin = 5)

In [77]:
fivedarray

array([[[[[1, 2, 3, 4]]]]])

In [79]:
fivedarray.ndim

5

In [80]:
fivedarray.shape

(1, 1, 1, 1, 4)

In [81]:
fivedarray.size

4

### Question
<li><b>Read the car_details.csv file using csv reader, convert them into lists of lists and then create a numpy array using the lists of list.</b></li>

In [63]:
from csv import reader
file = open('car_details.csv')
file_read = reader(file)
data = list(file_read)

In [64]:
print(data[1:6])

[['Maruti 800 AC', '2007', '60000', '70000', 'Petrol', 'Individual', 'Manual', 'First Owner'], ['Maruti Wagon R LXI Minor', '2007', '135000', '50000', 'Petrol', 'Individual', 'Manual', 'First Owner'], ['Hyundai Verna 1.6 SX', '2012', '600000', '100000', 'Diesel', 'Individual', 'Manual', 'First Owner'], ['Datsun RediGO T Option', '2017', '250000', '46000', 'Petrol', 'Individual', 'Manual', 'First Owner'], ['Honda Amaze VX i-DTEC', '2014', '450000', '141000', 'Diesel', 'Individual', 'Manual', 'Second Owner']]


In [65]:
csv_array = np.array(data[1:])
print(csv_array)

[['Maruti 800 AC' '2007' '60000' ... 'Individual' 'Manual' 'First Owner']
 ['Maruti Wagon R LXI Minor' '2007' '135000' ... 'Individual' 'Manual'
  'First Owner']
 ['Hyundai Verna 1.6 SX' '2012' '600000' ... 'Individual' 'Manual'
  'First Owner']
 ...
 ['Maruti 800 AC BSIII' '2009' '110000' ... 'Individual' 'Manual'
  'Second Owner']
 ['Hyundai Creta 1.6 CRDi SX Option' '2016' '865000' ... 'Individual'
  'Manual' 'First Owner']
 ['Renault KWID RXT' '2016' '225000' ... 'Individual' 'Manual'
  'First Owner']]


In [66]:
print(csv_array.ndim)

2


In [67]:
print(csv_array.shape)

(4340, 8)


In [68]:
print(csv_array.size)

34720


In [69]:
print(len(csv_array))

4340


## Special NumPy Arrays

<ol>
    <li><b>np.zeros():</b>The zeros() function is used to get a new array of given shape and type, filled with zeros.</li>
    <li><b>np.ones():</b>The ones() function is used to get a new array of given shape and type, filled with ones.</li>
</ol>

In [89]:
zeros_array = np.zeros(4)
print(zeros_array)
print(zeros_array.ndim)
print(zeros_array.shape)
print(zeros_array.size)

[0. 0. 0. 0.]
1
(4,)
4


In [91]:
zeros_array2d = np.zeros((3,5))
print(zeros_array2d)
print(zeros_array2d.ndim)
print(zeros_array2d.shape)
print(zeros_array2d.size)

[[0. 0. 0. 0. 0.]
 [0. 0. 0. 0. 0.]
 [0. 0. 0. 0. 0.]]
2
(3, 5)
15


In [92]:
zeros_array3d = np.zeros((3,5,2))
print(zeros_array3d)
print(zeros_array3d.ndim)
print(zeros_array3d.shape)
print(zeros_array3d.size)

[[[0. 0.]
  [0. 0.]
  [0. 0.]
  [0. 0.]
  [0. 0.]]

 [[0. 0.]
  [0. 0.]
  [0. 0.]
  [0. 0.]
  [0. 0.]]

 [[0. 0.]
  [0. 0.]
  [0. 0.]
  [0. 0.]
  [0. 0.]]]
3
(3, 5, 2)
30


In [93]:
zeros_array2d = np.zeros((3,5), dtype = 'int')
print(zeros_array2d)
print(zeros_array2d.ndim)
print(zeros_array2d.shape)
print(zeros_array2d.size)

[[0 0 0 0 0]
 [0 0 0 0 0]
 [0 0 0 0 0]]
2
(3, 5)
15


In [94]:
ones_array = np.ones(4)
print(ones_array)
print(ones_array.ndim)
print(ones_array.shape)
print(ones_array.size)
print(len(ones_array))

[1. 1. 1. 1.]
1
(4,)
4
4


In [95]:
ones_array_2d = np.ones((4, 5))
print(ones_array_2d)
print(ones_array_2d.ndim)
print(ones_array_2d.shape)
print(ones_array_2d.size)
print(len(ones_array_2d))

[[1. 1. 1. 1. 1.]
 [1. 1. 1. 1. 1.]
 [1. 1. 1. 1. 1.]
 [1. 1. 1. 1. 1.]]
2
(4, 5)
20
4


In [96]:
ones_array_2d = np.ones((4, 5), dtype = 'int')
print(ones_array_2d)
print(ones_array_2d.ndim)
print(ones_array_2d.shape)
print(ones_array_2d.size)
print(len(ones_array_2d))

[[1 1 1 1 1]
 [1 1 1 1 1]
 [1 1 1 1 1]
 [1 1 1 1 1]]
2
(4, 5)
20
4


<b>3. np.empty():</b>

<li>Return a new array of given shape and type, with random values.</li>
<li>Note : empty, unlike zeros, does not set the array values to zero, and may therefore be marginally faster.</li>
<li>It accepts three parameters. They are as follows:</li>
<ol>
    <li><b>shape:</b> Shape of the empty array, e.g., (2, 3) or 2.	Required</li>
    <li><b>dtype:</b> Desired output data-type for the array, e.g, numpy.int8. Default is numpy.float64.	optional</li>
    <li><b>order:</b> Whether to store multi-dimensional data in row-major (C-style) or column-major (Fortran-style) order in memory. optional</li>
</ol>

In [97]:
empty_array1d = np.empty(5)
print(empty_array1d)
print(empty_array1d.size)
print(empty_array1d.ndim)
print(empty_array1d.shape)
print(len(empty_array1d))

[2.12199579e-314 6.36598737e-314 1.06099790e-313 1.48539705e-313
 1.90979621e-313]
5
1
(5,)
5


In [98]:
empty_array2d = np.empty((2, 3))
print(empty_array2d)
print(empty_array2d.size)
print(empty_array2d.ndim)
print(empty_array2d.shape)
print(len(empty_array2d))

[[6.23042070e-307 4.67296746e-307 1.69121096e-306]
 [1.95821982e-306 1.89146896e-307 7.56571288e-307]]
6
2
(2, 3)
2


In [100]:
empty_array2d = np.empty((2, 3),dtype = 'int', order = "C")
print(empty_array2d)
print(empty_array2d.size)
print(empty_array2d.ndim)
print(empty_array2d.shape)
print(len(empty_array2d))

[[  508737360         372   511467904]
 [        372           0 -2147483648]]
6
2
(2, 3)
2


In [101]:
empty_array2d = np.empty((2, 3),dtype = 'int', order = "F")
print(empty_array2d)
print(empty_array2d.size)
print(empty_array2d.ndim)
print(empty_array2d.shape)
print(len(empty_array2d))

[[  508737360           0           1]
 [        372           0 -2147483648]]
6
2
(2, 3)
2


<b>4. np.arange()</b>
<li>This arange() function is similar like range in Python.</li>
<li>np.arange() returns an array with evenly spaced elements as per the interval. The interval mentioned is half-opened i.e. [Start, Stop)</li>
<li>It accepts four parameters which are as follows:</li>
<ol>
    <li>start : [optional] start of interval range. By default start = 0</li>
    <li>stop  : end of interval range</li>
    <li>step  : [optional] step size of interval. By default step size = 1,</li>  
    <li>dtype : type of output array</li>
</ol>

In [103]:
arrange_array = np.arange(1,11)
print(arrange_array)

[ 1  2  3  4  5  6  7  8  9 10]


In [104]:
arrange_array = np.arange(1,11, 3)
print(arrange_array)

[ 1  4  7 10]


In [105]:
arrange_array = np.arange(1,11, 3, 'float')
print(arrange_array)

[ 1.  4.  7. 10.]


<b>5.np.eye():</b>
<li>An array where diagonal element is filled with 1's.</li>
<li>It accepts four parameters which are as follows:</li>
<ol>
    <li>R: no of rows</li>
    <li>C: no of columns</li>
    <li>k : [int, optional, 0 by default]
          Diagonal we require; k>0 means diagonal above main diagonal or vice versa.</li>
    <li>dtype : [optional, float(by Default)] Data type of returned array.</li>


In [108]:
diagonal_matrix = np.eye(5,6,k = 1)
print(diagonal_matrix)

[[0. 1. 0. 0. 0. 0.]
 [0. 0. 1. 0. 0. 0.]
 [0. 0. 0. 1. 0. 0.]
 [0. 0. 0. 0. 1. 0.]
 [0. 0. 0. 0. 0. 1.]]


In [113]:
diagonal_matrix_2 = np.eye(5,6,k = -4)
print(diagonal_matrix_2)

[[0. 0. 0. 0. 0. 0.]
 [0. 0. 0. 0. 0. 0.]
 [0. 0. 0. 0. 0. 0.]
 [0. 0. 0. 0. 0. 0.]
 [1. 0. 0. 0. 0. 0.]]


In [114]:
diagonal_matrix_3 = np.eye(5,6,k = -4, dtype = 'int')
print(diagonal_matrix_3)

[[0 0 0 0 0 0]
 [0 0 0 0 0 0]
 [0 0 0 0 0 0]
 [0 0 0 0 0 0]
 [1 0 0 0 0 0]]


<b>6. np.linspace():</b> 
<li>It will create an array with values that are spaced linearly in a specified interval.</li>
<li>The numpy.linspace() function returns number spaces evenly w.r.t interval. Similar to numpy.arange() function but instead of step it uses sample number.</li>
<li>It accepts 5 parameters. They are as follows:</li>
<ol>
    <li>start  : [optional] start of interval range. By default start = 0</li>
    <li>stop   : end of interval range</li>
    <li>retstep : If True, return (samples, step). By default retstep = False</li>
    <li>num    : [int, optional] No. of samples to generate; default value is 50</li>
    <li>dtype  : type of output array</li>
</ol>

In [2]:
import numpy as np
arr1 = np.linspace(1, 32)

In [3]:
arr1

array([ 1.        ,  1.63265306,  2.26530612,  2.89795918,  3.53061224,
        4.16326531,  4.79591837,  5.42857143,  6.06122449,  6.69387755,
        7.32653061,  7.95918367,  8.59183673,  9.2244898 ,  9.85714286,
       10.48979592, 11.12244898, 11.75510204, 12.3877551 , 13.02040816,
       13.65306122, 14.28571429, 14.91836735, 15.55102041, 16.18367347,
       16.81632653, 17.44897959, 18.08163265, 18.71428571, 19.34693878,
       19.97959184, 20.6122449 , 21.24489796, 21.87755102, 22.51020408,
       23.14285714, 23.7755102 , 24.40816327, 25.04081633, 25.67346939,
       26.30612245, 26.93877551, 27.57142857, 28.20408163, 28.83673469,
       29.46938776, 30.10204082, 30.73469388, 31.36734694, 32.        ])

In [10]:
arr2 = np.linspace(1, 21, retstep = True, num = 6, dtype = 'int')
print(arr2)

(array([ 1,  5,  9, 13, 17, 21]), 4.0)


In [9]:
arr3 = np.linspace(1, 60, retstep = True, num = 10)
print(arr3)

(array([ 1.        ,  7.55555556, 14.11111111, 20.66666667, 27.22222222,
       33.77777778, 40.33333333, 46.88888889, 53.44444444, 60.        ]), 6.555555555555555)


## Create Numpy Arrays With Random Numbers

<li>Earlier, we read about some special functions with which we can create special numpy arras.</li>
<li>If we want our numpy arrays to have random values, then this can be achived by utilizing three special functions which is capable for generating random values.</li>
<li>They are:</li>
<ol>
    <li>rand()</li>
    <li>randn()</li>
    <li>randint()</li>
</ol>

<b>1. rand():</b>
<li>The numpy.random.rand() function creates an array of specified shape and fills it with random values.</li>
<li>This function is used to generate a random value between 0 and 1.</li>

**Syntax**
<code>
    np.random.rand(d0, d1, ..., dn)
</code>
<li>It returns the random values lying between 0 and 1 in a given shape.</li>

### 1D array

In [11]:
random1 = np.random.rand(4)
print(random1)
print(random1.ndim)
print(random1.shape)

[0.95463112 0.79351239 0.05302786 0.69651982]
1
(4,)


### 2D array

In [13]:
random2d = np.random.rand(2,3)
print(random2d)
print(random2d.ndim)
print(random2d.shape)

[[0.02887772 0.15483597 0.58150721]
 [0.12002761 0.30088594 0.47961064]]
2
(2, 3)


### 3D array

In [14]:
random3d = np.random.rand(2,3, 4)
print(random3d)
print(random3d.ndim)
print(random3d.shape)

[[[0.95752706 0.17805226 0.71027313 0.70360017]
  [0.725773   0.53440217 0.00397417 0.05835542]
  [0.38424546 0.55138782 0.67297395 0.86031444]]

 [[0.48158831 0.307162   0.28908663 0.14625126]
  [0.78863658 0.00911609 0.56076632 0.69301797]
  [0.80332935 0.97950198 0.08612256 0.79718827]]]
3
(2, 3, 4)


<b>2. randn():</b>
<li>The numpy.random.randn() function creates an array of specified shape and fills it with random values as per standard normal distribution.</li>
<li>This function is used to generate a random value close to zero(0).</li>
<li>This function may return positive as well as negative numbers as well.</li>


**Syntax**
<code>
    np.random.randn(d0, d1, ..., dn)
</code>


In [16]:
radnarray1 = np.random.randn(4)
print(radnarray1)
print(radnarray1.ndim)
print(radnarray1.shape)

[ 1.74789855 -0.87751881  1.28696745  1.5763005 ]
1
(4,)


In [17]:
radnarray2 = np.random.randn(4,5)
print(radnarray2)
print(radnarray2.ndim)
print(radnarray2.shape)

[[ 0.54986252 -0.16773915 -0.11622908  0.30182812  0.18112056]
 [ 2.6722416   1.87639397 -0.38419009  1.29041412  0.53937962]
 [ 1.93257508  0.63693674 -1.79050873 -0.00284375  1.19650053]
 [-0.54113153  0.61806044 -0.85315597  0.75394884  0.52727277]]
2
(4, 5)


In [18]:
radnarray3 = np.random.randn(4,5,2)
print(radnarray3)
print(radnarray3.ndim)
print(radnarray3.shape)

[[[-0.08755572  0.08033696]
  [-2.0364295   1.44252961]
  [ 0.07020553 -2.74748178]
  [ 0.03369992 -0.36103245]
  [-1.4717839   0.8165238 ]]

 [[ 0.09260011 -0.60485831]
  [ 1.5209407   0.51503588]
  [ 0.17628288  0.47206873]
  [ 1.06594324 -0.45132352]
  [-1.82448047 -0.81127458]]

 [[ 0.40173237 -0.05531126]
  [ 0.7642984  -0.42601103]
  [-1.21412479  1.43721162]
  [-0.09559734  1.09512413]
  [ 0.63453368 -1.21112776]]

 [[-0.62443027  0.17875261]
  [ 0.17546503 -0.61542145]
  [ 1.99503413 -1.35320111]
  [-0.42129052 -1.74956098]
  [-0.07132623  0.20879304]]]
3
(4, 5, 2)


<b>3. randint()</b>
<li>This function is used to generate random numbers between a given range.</li>
<li>randint() is an inbuilt function of the random module in Python3.</li>
<li>It accepts two parameters. They are:</li>
<ol>
    <li>start: a starting point which is an integer value.</li>
    <li>end: an ending point which is also an integer value.</li>
</ol>

**Syntax**
<code>
    random.randint(low, high=None, size=None, dtype=int)
</code>

In [29]:
randintarr = np.random.randint(5, 23)
print(randintarr)

19


In [30]:
randintarr2 = np.random.randint(5, 23, 6)
print(randintarr2)

[ 8 14  6 13 17 12]


In [33]:
randintarr3 = np.random.randint(5, 89, (2,3))
print(randintarr3)

[[75 67 44]
 [52 73 38]]


In [34]:
randintarr4 = np.random.randint(5, 89, (2,3,2))
print(randintarr4)

[[[60 10]
  [86 42]
  [21 58]]

 [[13 52]
  [43 59]
  [17 84]]]


## DataTypes In Numpy

<li>Numpy offers several datatypes to work with. Some of the datatypes are listed below:</li>

![](images/numpy_datatypes.png)

In [35]:
import warnings
warnings.filterwarnings('ignore')

nt8dtype = np.array([2,3,45,129,234,54], dtype = 'int8')
print(nt8dtype)

[   2    3   45 -127  -22   54]


In [36]:
nt16dtype = np.array([2,3,45,129,234,54], dtype = 'int16')
print(nt16dtype)

[  2   3  45 129 234  54]


In [ ]:
nt16dtype = np.array([2,3,45,129,234,54], dtype = 'int')
print(nt16dtype)

In [38]:
unt8dtype = np.array([2,3,45,129,234,540], dtype = 'uint8')
print(unt8dtype)

[  2   3  45 129 234  28]


In [39]:
256*2

512

In [45]:
float16dtype = np.array([2,3,45,129,238888874,545555], dtype = 'float16')
print(float16dtype)

[  2.   3.  45. 129.  inf  inf]


In [46]:
float32dtype = np.array([2,3,45,129,238888874,545555], dtype = 'float32')
print(float32dtype)

[2.0000000e+00 3.0000000e+00 4.5000000e+01 1.2900000e+02 2.3888888e+08
 5.4555500e+05]


In [47]:
float64dtype = np.array([2,3,45,129,23888887455555,54555555555555], dtype = 'float64')
print(float64dtype)

[2.00000000e+00 3.00000000e+00 4.50000000e+01 1.29000000e+02
 2.38888875e+13 5.45555556e+13]


In [53]:
complex1 = np.complex128([5+4j,3+2j, 4+4j])
print(complex1)

[5.+4.j 3.+2.j 4.+4.j]


In [54]:
complex1.dtype

dtype('complex128')

In [55]:
complex1.real

array([5., 3., 4.])

In [57]:
complex1.imag

array([4., 2., 4.])

In [58]:
np.conjugate(complex1)

array([5.-4.j, 3.-2.j, 4.-4.j])

In [63]:
a = np.arange('2022-12-01', '2022-12-31',2, dtype = 'datetime64')
print(a)

['2022-12-01' '2022-12-03' '2022-12-05' '2022-12-07' '2022-12-09'
 '2022-12-11' '2022-12-13' '2022-12-15' '2022-12-17' '2022-12-19'
 '2022-12-21' '2022-12-23' '2022-12-25' '2022-12-27' '2022-12-29']


In [64]:
print(a.dtype)

datetime64[D]


In [68]:
datetimedtype = np.datetime64('2022-12-23')
print(datetimedtype)
print(type(datetimedtype))

2022-12-23
<class 'numpy.datetime64'>


In [85]:
newdatetime = np.datetime64('2023-01-01')
print(newdatetime)
print(type(newdatetime))

2023-01-01
<class 'numpy.datetime64'>


In [86]:
days_left_df  = np.subtract(newdatetime, datetimedtype)
print(days_left_df)

9 days


In [88]:
print(type(days_left_df))
days_left_df.dtype

<class 'numpy.timedelta64'>


dtype('<m8[D]')

In [74]:
timedeltadtype = np.timedelta64('5','D')

In [82]:
adddatetimedelta = np.add(datetimedtype, timedeltadtype)
print(adddatetimedelta)
print(adddatetimedelta.dtype)

2022-12-28
datetime64[D]


In [83]:
subdatetimedelta = np.subtract(datetimedtype, timedeltadtype)
print(subdatetimedelta)
print(subdatetimedelta.dtype)

2022-12-18
datetime64[D]


In [90]:
stringdtype = np.array(['bimal', 'sherbahadur', 'solomon', 
                       'asmita'])
print(stringdtype)
print(stringdtype.dtype)

['bimal' 'sherbahadur' 'solomon' 'asmita']
<U11


<li>For more info about datetime datatype do visit this link.</li>
<li>https://numpy.org/doc/stable/reference/arrays.datetime.html</li>

## Arithmetic Operations In NumPy

<b>1. Addition:</b>
<li>We can add two numpy arrays of same shape. For adding two numpy arrays, we use <b>np.add(a,b)</b> function where a and b are two numpy arrays.</li>
<li>We can perform addition on any dimension arrays if they both are of same shape.</li>


In [94]:
array1 = np.array([[1,2,3], [4,5,6], [7,8,9]])
array2 = np.array([[9,8,7], [3,2,1], [6,5,4]])

In [97]:
addition1 = np.add(array1 , array2)
print(addition1)

[[10 10 10]
 [ 7  7  7]
 [13 13 13]]


In [96]:
array3 = np.array([1,2,3,4,5,6, 7,8,9])
array4 = np.array([9,8,7,3,2,1,6,5,4])

In [98]:
addition2 = np.add(array3, array4)
print(addition2)

[10 10 10  7  7  7 13 13 13]


In [99]:
array5 = np.random.randint(2, 45, (2,2,2))
array6 = np.random.randint(3, 13, (2,2,2))

In [100]:
print(array5)

[[[16 33]
  [ 7 29]]

 [[42 37]
  [22 28]]]


In [101]:
print(array6)

[[[ 4  9]
  [ 6 10]]

 [[ 5  4]
  [11  8]]]


In [102]:
sum_3d = np.add(array5, array6)
print(sum_3d)

[[[20 42]
  [13 39]]

 [[47 41]
  [33 36]]]


<b>2. Subtraction:</b>
<li>We can subtract two numpy arrays of same shape. For subtracting two numpy arrays, we use np.subtract(a,b) where a and b are two numpy arrays.</li>
<li>We can perform subtraction on any dimension arrays if they both are of same shape.</li>


In [104]:
array1 = np.random.randint(1, 8, 4)
array2 = np.random.randint(5, 12, 4)
print(array1)
print(array2)
sub12 = np.subtract(array1, array2)
print(sub12)

[4 3 2 5]
[ 7  8  9 10]
[-3 -5 -7 -5]


In [106]:
array3 = np.random.randint(5, 11, (2,4))
array4 = np.random.randint(1, 9, (2,4))
print(array3)
print()
print(array4)
print()
sub34 = np.subtract(array3, array4)
print(sub34)

[[ 5  8  7  8]
 [ 6  7 10  6]]

[[2 3 7 5]
 [2 5 8 6]]

[[3 5 0 3]
 [4 2 2 0]]


In [107]:
array5 = np.random.randint(5, 11, (3,3,3))
array6 = np.random.randint(1, 9, (3,3,3))
print(array5)
print()
print(array6)
print()
sub56 = np.subtract(array5, array6)
print(sub56)

[[[ 7  6  5]
  [10 10  8]
  [ 5 10  6]]

 [[10 10 10]
  [10  5  6]
  [10  8  7]]

 [[ 8  9 10]
  [ 7 10  8]
  [ 5  9  6]]]

[[[3 8 4]
  [4 6 5]
  [5 8 8]]

 [[6 3 8]
  [7 8 2]
  [6 1 1]]

 [[5 8 2]
  [4 3 8]
  [4 8 3]]]

[[[ 4 -2  1]
  [ 6  4  3]
  [ 0  2 -2]]

 [[ 4  7  2]
  [ 3 -3  4]
  [ 4  7  6]]

 [[ 3  1  8]
  [ 3  7  0]
  [ 1  1  3]]]


<b>3. Multiplication:</b>
<li>We can multiply two numpy arrays of same shape. For subtracting two numpy arrays, we use np.multiply(a,b) where a and b are two numpy arrays.</li>
<li>We can perform multiplication on any dimension arrays if they both are of same shape.</li>


In [108]:
array1 = np.random.randint(1, 8, 4)
array2 = np.random.randint(5, 12, 4)
print(array1)
print(array2)
mul12 = np.multiply(array1, array2)
print(mul12)

[6 2 3 5]
[ 7 10  8 11]
[42 20 24 55]


In [109]:
array3 = np.random.randint(5, 11, (2,4))
array4 = np.random.randint(1, 9, (2,4))
print(array3)
print()
print(array4)
print()
mul34 = np.multiply(array3, array4)
print(mul34)

[[ 9  8  5  6]
 [ 5  7 10  5]]

[[3 6 3 2]
 [1 7 4 4]]

[[27 48 15 12]
 [ 5 49 40 20]]


In [110]:
array5 = np.random.randint(5, 11, (2,2,2))
array6 = np.random.randint(1, 9, (2,2,2))
print(array5)
print()
print(array6)
print()
mul56 = np.multiply(array5, array6)
print(mul56)

[[[ 7  6]
  [ 5  9]]

 [[ 5  5]
  [10  5]]]

[[[8 5]
  [7 7]]

 [[7 2]
  [4 4]]]

[[[56 30]
  [35 63]]

 [[35 10]
  [40 20]]]


<b>4. Division:</b>
<li>We can divide two numpy arrays of same shape. For dividing two numpy arrays, we can use np.divide(a, b) where a and b are two numpy arrays.</li>
<li>We can perform division on any dimension arrays if they both are of same shape.</li>


In [111]:
array1 = np.random.randint(1, 8, 4)
array2 = np.random.randint(5, 12, 4)
print(array1)
print(array2)
div12 = np.divide(array1, array2)
print(div12)

[5 5 3 2]
[ 9 11  8  6]
[0.55555556 0.45454545 0.375      0.33333333]


In [112]:
array3 = np.random.randint(5, 11, (2,4))
array4 = np.random.randint(1, 9, (2,4))
print(array3)
print()
print(array4)
print()
div34 = np.divide(array3, array4)
print(div34)

[[8 6 9 6]
 [7 7 6 8]]

[[8 3 6 5]
 [4 8 2 6]]

[[1.         2.         1.5        1.2       ]
 [1.75       0.875      3.         1.33333333]]


In [113]:
array5 = np.random.randint(5, 11, (2,2,2))
array6 = np.random.randint(1, 9, (2,2,2))
print(array5)
print()
print(array6)
print()
div56 = np.divide(array5, array6)
print(div56)

[[[ 9 10]
  [ 7  6]]

 [[ 5  7]
  [ 7  8]]]

[[[5 4]
  [5 6]]

 [[5 8]
  [7 3]]]

[[[1.8        2.5       ]
  [1.4        1.        ]]

 [[1.         0.875     ]
  [1.         2.66666667]]]


<b>5. Modulus:</b>
<li>If a number is not exactly divisible by a divisor then the residuals is known as remainder.</li>
<li>Modulus also gives the remainder value of a number if that number is not excatly divisible by divisor.</li>
<li>If 54 is divided by 7, 54 is not excatly divisible by 7. Then in this scenario, 7 is the quotient and 5 is the remainder so is the modulus.</li>
<li>We can calculate the modulus of any two numpy arrays of same shape.</li>
<li>For calculating modulus, we use np.mod(a,b) where a and b are any two numpy arrays.</li>
<li>We can calculate the modulus on any dimension arrays if they both are of same shape.</li>

In [114]:
array1 = np.random.randint(2, 12, 4)
print(array1)
modulus1 = np.mod(array1, 3)
print()
print(modulus1)

[7 8 9 2]

[1 2 0 2]


In [115]:
array3 = np.random.randint(5, 11, (2,4))
array4 = np.random.randint(1, 9, (2,4))
print(array3)
print()
print(array4)
print()
mod34 = np.mod(array3, array4)
print(mod34)

[[ 5  5  9  8]
 [ 9 10 10 10]]

[[8 4 6 5]
 [1 8 8 7]]

[[5 1 3 3]
 [0 2 2 3]]


In [116]:
array5 = np.random.randint(5, 11, (2,2,2))
array6 = np.random.randint(1, 9, (2,2,2))
print(array5)
print()
print(array6)
print()
mod56 = np.mod(array5, array6)
print(mod56)

[[[ 9 10]
  [ 7  6]]

 [[ 9  9]
  [10 10]]]

[[[8 1]
  [4 7]]

 [[2 4]
  [2 4]]]

[[[1 0]
  [3 6]]

 [[1 1]
  [0 2]]]


<b>6. Power:</b>
<li>We can calculate square, cube and power of any two numbers present in an numpy array.</li>
<li>For calculating power, we use np.power(a,b) where a and b are two numpy arrays.</li>

In [117]:
array1 = np.random.randint(3, 9, 5)
print(array1)
print()
print(np.power(array1, 2))


[3 7 7 6 7]

[ 9 49 49 36 49]


In [118]:
array2 = np.random.randint(1, 15, (2,2))
print(array2)
print()
print(np.power(array2, 3))

[[4 3]
 [8 1]]

[[ 64  27]
 [512   1]]


In [119]:
array3 = np.random.randint(1, 9, (2,2,2))
print(array3)
print()
print(np.power(array3, 0.5))

[[[2 7]
  [7 3]]

 [[2 7]
  [1 4]]]

[[[1.41421356 2.64575131]
  [2.64575131 1.73205081]]

 [[1.41421356 2.64575131]
  [1.         2.        ]]]


<b>7. Reciprocal:</b>

<li>We can calculate the reciprocal of any given array of any dimension in numpy.</li>
<li>For calculating the reciprocal in numpy, we have np.reciprocal(a) where a is the numpy array.</li>

In [121]:
onedarray = np.random.randint(2, 8, 4)
print(onedarray)
print()
print(np.reciprocal(onedarray, dtype = 'float'))

[4 6 7 4]

[0.25       0.16666667 0.14285714 0.25      ]


In [122]:
twodarray = np.random.randint(2, 8, (2,2))
print(twodarray)
print()
print(np.reciprocal(twodarray, dtype = 'float'))

[[2 3]
 [6 4]]

[[0.5        0.33333333]
 [0.16666667 0.25      ]]


In [123]:
threedarray = np.random.randint(2, 8, (2,2,2))
print(threedarray)
print()
print(np.reciprocal(threedarray, dtype = 'float'))

[[[7 2]
  [6 4]]

 [[5 5]
  [7 5]]]

[[[0.14285714 0.5       ]
  [0.16666667 0.25      ]]

 [[0.2        0.2       ]
  [0.14285714 0.2       ]]]


## Statistical Functions In NumPy

<b>1. np.min(x):</b>
<li>We can find the minimum value present in any given array using np.min(x) where x is the given array.</li>
<li>In two dimension array, we can also find the minimum value along the axis using axis parameter in np.min(x, axis)</li>
<li>np.min(x, axis = 0) calculates the minimum value of an array across the columns.</li>
<li>np.min(x, axis = 1) calculates the minimum value of an array across the rows.</li>

<b>2. np.argmin(x):</b>
<li>We can find the index of the minimum value present in any given array using np.argmin(x) where x is the given array.</li>
<li>In two dimension array, we can also find the index of the minimum value along the axis using axis parameter in np.argmin(x, axis)</li>
<li>np.argmin(x, axis = 0) calculates the minimum index value of an array across the columns.</li>
<li>np.argmin(x, axis = 1) calculates the minimum index value of an array across the rows.</li>


<b>3. np.max(x):</b>
<li>We can find the maximum value present in any given array using np.max(x) where x is the given array.</li>
<li>In two dimension array, we can also find the maximum value along the axis using axis parameter in np.max(x, axis)</li>
<li>np.max(x, axis = 0) calculates the maximum value of an array across the columns.</li>
<li>np.max(x, axis = 1) calculates the maximum value of an array across the rows.</li>


<b>4. np.argmax(x):</b>
<li>We can find the index of the maximum value present in any given array using np.argmax(x) where x is the given array.</li>
<li>In two dimension array, we can also find the index of the maximum value along the axis using axis parameter in np.argmax(x, axis)</li>
<li>np.argmax(x, axis = 0) calculates the maximum index value of an array across the columns.</li>
<li>np.argmax(x, axis = 1) calculates the maximum index value of an array across the rows.</li>


<b>5. np.sqrt(x):</b>
<li>We can find the square root of any value present in any given array using np.sqrt(x) where x is the given array.</li>

<b>6. np.sin(x):</b>
<li>We can find the sine value of values present in any given array using np.sin(x) where x is the given array.</li>

<b>7. np.cos(x):</b>
<li>We can find the cos value of values present in any given array using np.cos(x) where x is the given array.</li>

<b>8. np.cumsum(x):</b>
<li>We can find the cumulative sum of any given array using np.cumsum(x) where x is the given array.</li>
<li>In two dimension array, we can also find the cumulative sum along the axis using axis parameter in np.cumsum(x, axis)</li>
<li>np.cumsum(x, axis = 0) calculates the cumulative sum of an array across the columns.</li>
<li>np.cumsum(x, axis = 1) calculates the cumulative sum of an array across the rows.</li>


### np.min()

In [2]:
import numpy as np
arr1 = np.array([1, 2,3,4,5,6,7,8])
print(np.min(arr1))

1


In [3]:
arr2 = np.array([[1,2,3,4], [5,6,7,8]])
print(arr2.ndim)

2


In [4]:
print(np.min(arr2))

1


In [5]:
print(np.min(arr2, axis = 0))

[1 2 3 4]


In [6]:
print(np.min(arr2, axis = 1))

[1 5]


In [7]:
arr3 = np.array([[[1,2], [3,4]], [[5,6], [7,8]]])
print(arr3)

[[[1 2]
  [3 4]]

 [[5 6]
  [7 8]]]


In [8]:
np.min(arr3)

1

In [9]:
np.min(arr3, axis = 0)

array([[1, 2],
       [3, 4]])

In [10]:
np.min(arr3, axis = 1)

array([[1, 2],
       [5, 6]])

In [11]:
np.min(arr3, axis = 2)

array([[1, 3],
       [5, 7]])

#### np.argmin()

In [12]:
arr1 = np.array([9,5,2,3,8])

In [13]:
print(np.argmin(arr1))

2


In [31]:
arr2 = np.array([[5,2],[4,3]])

In [32]:
print(arr2)

[[5 2]
 [4 3]]


In [33]:
print(np.argmin(arr2, axis = 0))

[1 0]


In [17]:
print(np.argmin(arr2, axis = 1))

[1 0]


In [22]:
arr3 = np.array([[[4,2], [6,3]], [[5,8], [1,7]]])
print(arr3)
print(arr3.ndim)

[[[4 2]
  [6 3]]

 [[5 8]
  [1 7]]]
3


In [23]:
print(np.argmin(arr3, axis = 0))

[[0 0]
 [1 0]]


### np.max()

In [34]:
arr1 = np.array([9,5,2,3,8])
print(np.max(arr1))

9


In [35]:
arr2 = np.array([[5,2],[4,3]])
print(np.max(arr2))

5


In [36]:
print(np.max(arr2, axis = 0))

[5 3]


In [37]:
print(np.max(arr2, axis = 1))

[5 4]


In [38]:
arr3 = np.array([[[4,2], [6,3]], [[5,8], [1,7]]])
print(np.max(arr3))

8


In [40]:
np.max(arr3, axis = 0)

array([[5, 8],
       [6, 7]])

In [41]:
np.max(arr3, axis = 1)

array([[6, 3],
       [5, 8]])

In [42]:
np.max(arr3, axis = 2)

array([[4, 6],
       [8, 7]])

#### np.argmax()

In [43]:
arr1 = np.array([9,5,2,3,8])
print(np.argmax(arr1))



0


In [44]:
arr2 = np.array([[5,2],[4,3]])
print(np.argmax(arr2))

0


In [45]:
print(np.argmax(arr2, axis = 0))

[0 1]


In [46]:
print(np.argmax(arr2, axis = 1))

[0 0]


In [47]:
arr3 = np.array([[[4,2], [6,3]], [[5,8], [1,7]]])
print(np.argmax(arr3))

5


In [48]:
print(np.argmax(arr3, axis = 0))

[[1 1]
 [0 1]]


In [49]:
print(np.argmax(arr3, axis = 1))

[[1 1]
 [0 0]]


In [50]:
print(np.argmax(arr3, axis = 2))

[[0 0]
 [1 1]]


#### np.sqrt()

In [51]:
arr1 = np.array([1,4,9,16,25,36])
np.sqrt(arr1)

array([1., 2., 3., 4., 5., 6.])

In [52]:
arr2 = np.array([[1,4,9],[16,25,36]])
np.sqrt(arr2)

array([[1., 2., 3.],
       [4., 5., 6.]])

In [53]:
arr3 = np.array([[[1,4,9],[16,25,36]], [[49, 64, 81], [100, 121, 144]]])
print(arr3)
print(arr3.ndim)

[[[  1   4   9]
  [ 16  25  36]]

 [[ 49  64  81]
  [100 121 144]]]
3


In [54]:
np.sqrt(arr3)

array([[[ 1.,  2.,  3.],
        [ 4.,  5.,  6.]],

       [[ 7.,  8.,  9.],
        [10., 11., 12.]]])

#### np.sin()

In [60]:
np.sin(np.deg2rad([0,30,45,60,90, 180]))

array([0.00000000e+00, 5.00000000e-01, 7.07106781e-01, 8.66025404e-01,
       1.00000000e+00, 1.22464680e-16])

#### np.cos()

In [61]:
np.cos(np.deg2rad([0,30,45,60,90, 180]))

array([ 1.00000000e+00,  8.66025404e-01,  7.07106781e-01,  5.00000000e-01,
        6.12323400e-17, -1.00000000e+00])

### np.cumsum()

In [62]:
arr1 = np.array([9,5,2,3,8])
print(np.cumsum(arr1))


[ 9 14 16 19 27]


In [63]:
arr2 = np.array([[5,2],[4,3]])
print(np.cumsum(arr2))

[ 5  7 11 14]


In [64]:
arr2 = np.array([[5,2],[4,3]])
print(np.cumsum(arr2, axis = 0))

[[5 2]
 [9 5]]


In [65]:
arr2 = np.array([[5,2],[4,3]])
print(np.cumsum(arr2, axis = 1))

[[5 7]
 [4 7]]


In [66]:
arr3 = np.array([[[4,2], [6,3]], [[5,8], [1,7]]])
print(np.cumsum(arr3))

[ 4  6 12 15 20 28 29 36]


In [67]:
print(arr3)

[[[4 2]
  [6 3]]

 [[5 8]
  [1 7]]]


In [68]:
np.cumsum(arr3, axis = 0)

array([[[ 4,  2],
        [ 6,  3]],

       [[ 9, 10],
        [ 7, 10]]])

In [69]:
np.cumsum(arr3, axis = 1)

array([[[ 4,  2],
        [10,  5]],

       [[ 5,  8],
        [ 6, 15]]])

In [70]:
np.cumsum(arr3, axis = 2)

array([[[ 4,  6],
        [ 6,  9]],

       [[ 5, 13],
        [ 1,  8]]])

## Shape & Reshape In Numpy Arrays

<li>The shape gives a tuple of array dimensions and can be used to change the dimensions of an array.</li>
<li>If we form a one dimension array of size 4 then the shape of that array is (4,).</li>
<li>If we have a two dimension array where there are 4 rows and 3 columns then the shape of the array is (4, 3).</li>
<li>The reshape function gives a new shape to an array without changing its data.</li>
<li>It creates a new array and does not modify the original array itself.</li>
<li>Suppose, we have a one dimensional array of size 12 then we can reshape that array as (2,6), (3,4), (4,3) and (6,2) new numpy arrays of dimension 2.</li>
<li>We can check the shape of an array using arr.shape whereas we can reshape an numpy array to another shape using np.reshape(arr, new_shape).</li>



In [71]:
array1 = np.array([1,2,3,4,5,6,7,8,9,10,11,12])
print(array1)
print(array1.ndim)
print(array1.shape)

[ 1  2  3  4  5  6  7  8  9 10 11 12]
1
(12,)


In [72]:
array2 = np.reshape(array1, (12,1))
print(array2.shape)
print(array2)

(12, 1)
[[ 1]
 [ 2]
 [ 3]
 [ 4]
 [ 5]
 [ 6]
 [ 7]
 [ 8]
 [ 9]
 [10]
 [11]
 [12]]


In [73]:
array3 = np.reshape(array1, (2, 6))
print(array3.shape)
print(array3)

(2, 6)
[[ 1  2  3  4  5  6]
 [ 7  8  9 10 11 12]]


In [74]:
array4 = np.reshape(array1, (4, 3))
print(array4.shape)
print(array4)

(4, 3)
[[ 1  2  3]
 [ 4  5  6]
 [ 7  8  9]
 [10 11 12]]


## BroadCasting In NumPy Arrays
<li>We all know that we can easily perform any arithmetic operations on two arrays of same dimensions or shape.</li>
<li>But we can also perform arithmetic operations on two arrays if they donot have same dimensions as well.</li>
<li>The term broadcasting refers to how numpy treats arrays with different Dimension during arithmetic operations.</li>
<li>The smaller array is broadcast across the larger array so that they have compatible shapes.</li>
<li>But there are some rules that two arrays must follow inorder to be broadcasted.</li>

### Broadcasting Rules:

**Broadcasting two arrays together follow these rules:**

<li>If the arrays don’t have same shape then prepend 1s to the shape of lower rank array until they have same length.</li>
<li>The arrays can be broadcast together if they are compatible with all dimensions.</li>
<li>The two arrays are compatible in a dimension if they have the same size in the dimension.</li>
<li>The two arrays are also compatible in a dimension if one of the arrays has size 1 in that dimension.</li>
<li>The broadcasted array behaves as if it had shape equal to the element-wise maximum of shapes of the two input arrays.</li>
<li>In any dimension where one array had size 1 and the other array had size greater than 1, the first array behaves as if it were copied along that dimension.</li>

### Checking Compatibility For BroadCasting

#### Example1

In [ ]:
x.shape == (2,3)

y.shape == (2,3)
y.shape == (3,2)
y.shape == (1,3)
y.shape == (3,1)
y.shape == (3,)
y.shape == (2,1)

In [1]:
import numpy as np

In [14]:
x = np.random.rand(2,3)

In [17]:
y = np.random.rand(2, 1)

In [18]:
x + y

array([[1.23570218, 1.4947771 , 1.16518451],
       [1.22590882, 0.52629996, 0.88960987]])

#### Example2

In [ ]:
x.shape == (1000,256,256,256)

y.shape == (1000, 256, 256, 256)
y.shape == (1, 256, 256, 1)
y.shape == (256,256,256)
y.shape == (1000, 1, 256, 1)
y.shape == (1000,256,256)
y.shape == (256,1000,256,256)
y.shape == (256,256,256,1000)

#### Example 3

In [ ]:

x.shape == (1,2,3,5,1,11,1,17)
y.shape == (1,7,1,1,17)


x.shape == (1,2,3,4,5,6,7,1,1)
y.shape == (4,5,1,7,3,2)

## Indexing & Slicing In NumPy Arrays

### Indexing
<li>Numpy indexing is used for accessing an element from an array by giving it an index value that starts from 0.</li>
<li>We can also use negative indexing in NumPy arrays like in lists and strings.</li>
<li>Negative indexing starts from the end of the array. It starts from a negative index value, i.e. -1.</li>

### Slicing
<li>Slicing NumPy arrays means extracting elements from an array in a specific range.</li>
<li>Slicing in python means taking elements from one given index to another given index.</li>
<li>We pass slice instead of index like this: [start:end].</li>
<li>We can also define the step, like this: [start:end:step].</li>
<li>If we don't pass start it is considered 0.</li>
<li>If we don't pass end it is considered as length of array in that dimension.</li>
<li>If we don't pass step it is considered 1.</li>




### Indexing in 1D

In [2]:
import numpy as np
arr1 = np.array([5,6,8,2,9,4])
print(arr1)

[5 6 8 2 9 4]


In [3]:
arr1[2]

8

In [4]:
arr1[[2,3,4]]

array([8, 2, 9])

### Indexing in 2D

In [5]:
arr2d = np.random.randint(3, 9, (2,2))
print(arr2d)

[[8 3]
 [5 6]]


In [6]:
arr2d[0][1]

3

In [7]:
arr2d[1][0]

5

### Indexing in 3D

In [8]:
arr3d = np.random.randint(1, 21, (3,3,3))
print(arr3d)

[[[12  5  1]
  [16  6  3]
  [ 4  3 16]]

 [[15 20 11]
  [13  8 11]
  [ 3 15  7]]

 [[11 14 14]
  [ 8  8  7]
  [11 14 15]]]


In [23]:
arr3d[1,1,1]

8

In [10]:
arr3d[2][0][2]

14

In [ ]:
arr3d[3][1]

### Slicing In 1D

In [12]:
arr1

array([5, 6, 8, 2, 9, 4])

In [13]:
arr1[2:5]

array([8, 2, 9])

In [14]:
arr1[-3:]

array([2, 9, 4])

In [15]:
arr1[:3]

array([5, 6, 8])

### Slicing In 2D

In [16]:
arr2d = np.random.randint(1,20,(4,4))
print(arr2d)

[[ 3 17  4 18]
 [10 10 16 13]
 [17 15 13 15]
 [ 7 11 16  1]]


In [26]:
arr2d[0::2, 0::3]

array([[ 3, 18],
       [17, 15]])

In [22]:
arr2d[:3, 1:3]

array([[17,  4],
       [10, 16],
       [15, 13]])

In [24]:
arr2d[:2,1::2]

array([[17, 18],
       [10, 13]])

### Slicing In 3D

In [30]:
arr3d = np.random.randint(1, 50, (4,4,4))
print(arr3d)

[[[25 40 32  3]
  [34 44 16  3]
  [16 37 38 22]
  [31 49 42 27]]

 [[23 27 17 10]
  [ 5 30  9 41]
  [ 3  4  2 47]
  [14 12 35 40]]

 [[41 37 49 28]
  [15  1 19 28]
  [33 28 10 16]
  [ 8 31 19 33]]

 [[10 10  3 27]
  [45 17 20 10]
  [48 12  5 16]
  [47 40 44 15]]]


In [ ]:
arr3d[:,1:2,::2]

In [31]:
arr3d[:2, 1::2,0:3:2]

array([[[34, 16],
        [31, 42]],

       [[ 5,  9],
        [14, 35]]])

### Types Of Indexing

<li>There are two types of indexing. They are :</li>
<ol>
    <b><li>Basic Slicing & Indexing</li></b>
    <b><li>Advanced Indexing</li></b>
</ol>
<li>Advanced Indexing is further divided into two types. They are:</li>
<ol>
    <b><li>Purely integer indexing</li></b>
    <b><li>Boolean indexing.</li></b>
</ol>

In [32]:
a = np.array([4,6,9,12,8])

In [34]:
a[a>8]

array([ 9, 12])

In [35]:
twodarr = np.random.randint(1,90, (3,3))
twodarr

array([[69, 14, 29],
       [37,  2, 84],
       [68, 12, 19]])

In [37]:
twodarr[twodarr < 20]

array([14,  2, 12, 19])

#### Question
<li>Load <b>car_details.csv</b> file into list of list using csv reader</li> 
<li>Then create a 2d numpy arrray from the list of list.</li>
<li>Extract the <b>first five rows</b> of the numpy array using slicing.</li>
<li>Extract the <b>last five rows</b> of the numpy array using slicing.</li>
<li>Check the <b>shape</b> and <b>dimension</b> of the dataset.</li>
<li>Find some compatible shapes in which we can <b>reshape</b> it.</li>
<li>If we have another numpy array of shape (8, 4340). Can we perform any operations on them? Is it <b>broadcastable</b>?</li>
<li>Find the <b>selling_prices</b> of all cars.</li>
<li>Find the <b>km_driven</b> for all cars.</li>
<li>How can we perform <b>boolean indexing</b> in the given data.</li>

In [4]:
import numpy as np

In [5]:
from csv import reader
file = open('car_details.csv')
file_read = reader(file)
data = list(file_read)
print(data[:5])

[['name', 'year', 'selling_price', 'km_driven', 'fuel', 'seller_type', 'transmission', 'owner'], ['Maruti 800 AC', '2007', '60000', '70000', 'Petrol', 'Individual', 'Manual', 'First Owner'], ['Maruti Wagon R LXI Minor', '2007', '135000', '50000', 'Petrol', 'Individual', 'Manual', 'First Owner'], ['Hyundai Verna 1.6 SX', '2012', '600000', '100000', 'Diesel', 'Individual', 'Manual', 'First Owner'], ['Datsun RediGO T Option', '2017', '250000', '46000', 'Petrol', 'Individual', 'Manual', 'First Owner']]


In [6]:
twodarray = np.array(data[1:])
print(twodarray)
print(twodarray.shape)
print(twodarray.ndim)

[['Maruti 800 AC' '2007' '60000' ... 'Individual' 'Manual' 'First Owner']
 ['Maruti Wagon R LXI Minor' '2007' '135000' ... 'Individual' 'Manual'
  'First Owner']
 ['Hyundai Verna 1.6 SX' '2012' '600000' ... 'Individual' 'Manual'
  'First Owner']
 ...
 ['Maruti 800 AC BSIII' '2009' '110000' ... 'Individual' 'Manual'
  'Second Owner']
 ['Hyundai Creta 1.6 CRDi SX Option' '2016' '865000' ... 'Individual'
  'Manual' 'First Owner']
 ['Renault KWID RXT' '2016' '225000' ... 'Individual' 'Manual'
  'First Owner']]
(4340, 8)
2


In [7]:
for item in twodarray:
    item[2] = int(item[2])
    item[3] = int(item[3])
    break

In [8]:
print(twodarray)

[['Maruti 800 AC' '2007' '60000' ... 'Individual' 'Manual' 'First Owner']
 ['Maruti Wagon R LXI Minor' '2007' '135000' ... 'Individual' 'Manual'
  'First Owner']
 ['Hyundai Verna 1.6 SX' '2012' '600000' ... 'Individual' 'Manual'
  'First Owner']
 ...
 ['Maruti 800 AC BSIII' '2009' '110000' ... 'Individual' 'Manual'
  'Second Owner']
 ['Hyundai Creta 1.6 CRDi SX Option' '2016' '865000' ... 'Individual'
  'Manual' 'First Owner']
 ['Renault KWID RXT' '2016' '225000' ... 'Individual' 'Manual'
  'First Owner']]


In [9]:
twodarray[:5]

array([['Maruti 800 AC', '2007', '60000', '70000', 'Petrol',
        'Individual', 'Manual', 'First Owner'],
       ['Maruti Wagon R LXI Minor', '2007', '135000', '50000', 'Petrol',
        'Individual', 'Manual', 'First Owner'],
       ['Hyundai Verna 1.6 SX', '2012', '600000', '100000', 'Diesel',
        'Individual', 'Manual', 'First Owner'],
       ['Datsun RediGO T Option', '2017', '250000', '46000', 'Petrol',
        'Individual', 'Manual', 'First Owner'],
       ['Honda Amaze VX i-DTEC', '2014', '450000', '141000', 'Diesel',
        'Individual', 'Manual', 'Second Owner']], dtype='<U58')

In [10]:
twodarray[-5:]

array([['Hyundai i20 Magna 1.4 CRDi (Diesel)', '2014', '409999', '80000',
        'Diesel', 'Individual', 'Manual', 'Second Owner'],
       ['Hyundai i20 Magna 1.4 CRDi', '2014', '409999', '80000',
        'Diesel', 'Individual', 'Manual', 'Second Owner'],
       ['Maruti 800 AC BSIII', '2009', '110000', '83000', 'Petrol',
        'Individual', 'Manual', 'Second Owner'],
       ['Hyundai Creta 1.6 CRDi SX Option', '2016', '865000', '90000',
        'Diesel', 'Individual', 'Manual', 'First Owner'],
       ['Renault KWID RXT', '2016', '225000', '40000', 'Petrol',
        'Individual', 'Manual', 'First Owner']], dtype='<U58')

In [11]:
twodarray.size

34720

<b><li>(10,3472), (5,6944), (4,8640), (2,17360) are the other shapes where we can reshape our twodarray.</li></b>

In [12]:
prices = twodarray[:,2]
print(prices)

['60000' '135000' '600000' ... '110000' '865000' '225000']


In [13]:
km_driven = twodarray[:,3]

In [14]:
prices[prices < '50000']

array(['135000', '250000', '450000', ..., '409999', '110000', '225000'],
      dtype='<U58')

In [16]:
twodarray[twodarray[:,2].astype(np.int32) < 50000]

array([['Maruti 800 Std', '1998', '40000', '40000', 'Petrol',
        'Individual', 'Manual', 'Fourth & Above Owner'],
       ['Maruti 800 AC BSII', '2001', '43000', '100000', 'Petrol',
        'Individual', 'Manual', 'Second Owner'],
       ['Tata Nano Std', '2011', '40000', '19000', 'Petrol',
        'Individual', 'Manual', 'First Owner'],
       ['Maruti 800 EX', '2001', '40000', '30000', 'Petrol',
        'Individual', 'Manual', 'First Owner'],
       ['Tata Nano Cx BSIII', '2014', '45000', '7000', 'Petrol',
        'Individual', 'Manual', 'Second Owner'],
       ['Maruti 800 AC BSII', '2001', '45000', '72539', 'Petrol',
        'Individual', 'Manual', 'Second Owner'],
       ['Tata Indica DLE', '2006', '45000', '120000', 'Diesel',
        'Individual', 'Manual', 'Second Owner'],
       ['Tata Nano LX SE', '2012', '35000', '35000', 'Petrol',
        'Individual', 'Manual', 'Third Owner'],
       ['Maruti 800 EX', '2004', '30000', '60000', 'Petrol',
        'Individual', 'Manual', '

## Iterating NumPy Arrays
<li>Iterating means going through elements one by one.</li>
<li>As we deal with multi-dimensional arrays in numpy, we can do this using basic for loop of python.</li>
<li>We can iterate over the items of a numpy array by using a single for loop for 1D array.</li>
<li>If we are working with 2D arrays, then we can access the items of 2d numpy array by iterating over 2 times.</li>
<li>If we are working with 3D arrays then we can access the items of a 3d numpy array by iterating over 3 times.</li>
<li>In basic for loops, iterating through each scalar of an array we need to use n for loops which can be difficult to write for arrays with very high dimensionality.</li>
<li>In numpy, we have <b>np.nditer()</b> which is a helping function that can be used from very basic to very advanced iterations.</li>
<li>It solves some basic issues which we face in iteration.</li>
<li>If we want to access both the index and the items of n-d array, then we can use <b>np.ndenumerate()</b> function.</li>


In [17]:
arr = np.array([1,2,3,4,5])
for item in arr:
    print(item)

1
2
3
4
5


In [18]:
arr2d = np.random.randint(1, 12, (3,3))
print(arr2d)

[[ 7  1  1]
 [ 8 11  6]
 [ 5 10  7]]


In [20]:
for item in arr2d:
    for subitem in item:
        print(subitem)


7
1
1
8
11
6
5
10
7


In [21]:
arr3d = np.random.randint(1, 29, (2,2,2))
print(arr3d)

[[[18  3]
  [ 4 22]]

 [[ 1 20]
  [24 13]]]


In [24]:
for item in arr3d:
    for subitem in item:
        for sub in subitem:
            print(sub)


18
3
4
22
1
20
24
13


In [28]:
for item in np.nditer(arr):
    print(item)

1
2
3
4
5


In [29]:
for item in np.nditer(arr2d):
    print(item)

7
1
1
8
11
6
5
10
7


In [30]:
for item in np.nditer(arr3d):
    print(item)

18
3
4
22
1
20
24
13


In [31]:
for idx, item in np.ndenumerate(arr):
    print(idx, item)


(0,) 1
(1,) 2
(2,) 3
(3,) 4
(4,) 5


In [32]:
for idx, item in np.ndenumerate(arr2d):
    print(idx, item)

(0, 0) 7
(0, 1) 1
(0, 2) 1
(1, 0) 8
(1, 1) 11
(1, 2) 6
(2, 0) 5
(2, 1) 10
(2, 2) 7


In [33]:
for idx, item in np.ndenumerate(arr3d):
    print(idx, item)

(0, 0, 0) 18
(0, 0, 1) 3
(0, 1, 0) 4
(0, 1, 1) 22
(1, 0, 0) 1
(1, 0, 1) 20
(1, 1, 0) 24
(1, 1, 1) 13


## Copy Vs View In Numpy Arrays:

![](images/copy_vs_view.png)

<li>We can use <b>np.copy()</b> function to create a new copy of the existing numpy array.</li>
<li>We can use <b>x.view()</b> function to create a view of an existing numpy array where x is an existing numpy array.</li>

In [34]:
arr1 = np.array([1,2,3,4])
print(arr1)

[1 2 3 4]


In [35]:
arr2 = np.copy(arr1)
print(arr2)

[1 2 3 4]


In [37]:
arr3 = arr1.view()
print(arr3)

[1 2 3 4]


In [38]:
arr2[1] = 20

In [39]:
print(arr2)

[ 1 20  3  4]


In [40]:
print(arr1)

[1 2 3 4]


In [41]:
arr3[1] = 30

In [42]:
print(arr3)

[ 1 30  3  4]


In [43]:
print(arr1)

[ 1 30  3  4]


In [44]:
arr1[1] = 12

In [45]:
print(arr1)

[ 1 12  3  4]


In [46]:
print(arr2)

[ 1 20  3  4]


In [47]:
print(arr3)

[ 1 12  3  4]


## Joining NumPy Arrays
<li>Joining means putting contents of two or more arrays in a single array.</li>
<li>In SQL we join tables based on a key, whereas in NumPy we join arrays by axes.</li>
<li>NumPy provides various functions to combine arrays.</li>
<li>Some of the functions that can be used to combined two numpy arrays are as follows:</li>
<ol>
    <b><li>numpy.concatenate()</li></b>
    <b><li>numpy.stack()</li></b>
    <b><li>numpy.block()</li></b>
</ol>


### 1. numpy.concatenate():

<li>The concatenate() function in NumPy joins two or more arrays along a specified axis. </li>
<li>If axis is not explicitly specified, it is taken as 0.</li>

**Syntax**
<code>
    numpy.concatenate((array1, array2, ...), axis=0)
</code>

<li>The first argument is a tuple of arrays we intend to join.</li>
<li>The second argument is the axis along which we need to join these arrays.</li>


In [50]:
onedarray1 = np.array([1,3,5,7])
print(onedarray1)
onedarray2 = np.array([2,4,6])
print()
print(onedarray2)

[1 3 5 7]

[2 4 6]


In [51]:
concat1d = np.concatenate([onedarray1, onedarray2])
print(concat1d)

[1 3 5 7 2 4 6]


In [53]:
twodarray1 = np.random.randint(1, 19, (2,2))
twodarray2 = np.random.randint(21, 39, (2,2))

print(twodarray1)
print(twodarray1.shape)
print()
print(twodarray2)
print(twodarray2.shape)

[[ 1  1]
 [12  9]]
(2, 2)

[[30 22]
 [36 36]]
(2, 2)


In [57]:
concat2d = np.concatenate([twodarray1, twodarray2])
print(concat2d)
print(concat2d.shape)

[[ 1  1]
 [12  9]
 [30 22]
 [36 36]]
(4, 2)


In [58]:
concat2d = np.concatenate([twodarray1, twodarray2], axis = 1)
print(concat2d)
print(concat2d.shape)

[[ 1  1 30 22]
 [12  9 36 36]]
(2, 4)


In [59]:
arr2d = np.random.randint(1,12, (3,3))
print(arr2d)

[[ 1  7  6]
 [ 2  3 11]
 [ 5  8  5]]


In [60]:
column_vector = np.array([[1,5,7]])

In [61]:
arr2d_concat = np.concatenate([arr2d, column_vector], axis = 0)
print(arr2d_concat)
print(arr2d_concat.shape)

[[ 1  7  6]
 [ 2  3 11]
 [ 5  8  5]
 [ 1  5  7]]
(4, 3)


In [63]:
column_vector = np.array([[1],[5], [7]])
print(column_vector)

[[1]
 [5]
 [7]]


In [64]:
arr2d_concat = np.concatenate([arr2d, column_vector], axis = 1)
print(arr2d_concat)

[[ 1  7  6  1]
 [ 2  3 11  5]
 [ 5  8  5  7]]


In [65]:
arr3d1 = np.random.randint(1,21,(2,2,2))
print(arr3d1)


arr3d2 = np.random.randint(31, 60, (2,2,2))
print()
print(arr3d2)

[[[16  6]
  [10 19]]

 [[ 5 13]
  [19  1]]]

[[[50 34]
  [52 53]]

 [[34 59]
  [48 42]]]


In [68]:
arr_concat0 = np.concatenate([arr3d1, arr3d2], axis= 0)
print(arr_concat0)
print(arr_concat0.shape)

[[[16  6]
  [10 19]]

 [[ 5 13]
  [19  1]]

 [[50 34]
  [52 53]]

 [[34 59]
  [48 42]]]
(4, 2, 2)


In [69]:
arr_concat1 = np.concatenate([arr3d1, arr3d2], axis = 1)
print(arr_concat1)

[[[16  6]
  [10 19]
  [50 34]
  [52 53]]

 [[ 5 13]
  [19  1]
  [34 59]
  [48 42]]]


In [70]:
arr_concat2 = np.concatenate([arr3d1, arr3d2], axis = 2)
print(arr_concat2)

[[[16  6 50 34]
  [10 19 52 53]]

 [[ 5 13 34 59]
  [19  1 48 42]]]


## 2. numpy.stack():

<li>Stacking is same as concatenation, the only difference is that stacking is done along a new axis.</li>
<li>stack() creates a new array which has 1 more dimension than the input arrays. If we stack 2 1-D arrays, the resultant array will have 2 dimensions.</li>
<li>We can concatenate two 1-D arrays along the second axis which would result in putting them one over the other, ie. stacking.</li>
<li>We pass a sequence of arrays that we want to join to the <b>np.stack()</b> method along with the axis. If axis is not explicitly passed it is taken as 0.</li>
<li>If we want to join 2 arrays using stacking operation, then they must have the same shape and dimensions. (e.g. both (2,3)–> 2 rows,3 columns)</li>


In [4]:
import numpy as np
arr1d1 = np.random.randint(1, 21, 5)
print(arr1d1)

[16 19  7  8 16]


In [5]:
arr1d2 = np.random.randint(22, 41, 5)
print(arr1d2)

[29 28 34 27 23]


In [9]:
stack1d1 = np.stack([arr1d1, arr1d2], axis = 0)
print(stack1d1)

[[16 19  7  8 16]
 [29 28 34 27 23]]


In [11]:
np.concatenate([arr1d1, arr1d2])

array([16, 19,  7,  8, 16, 29, 28, 34, 27, 23])

In [12]:
stack1d2 = np.stack([arr1d1, arr1d2], axis = 1)
print(stack1d2)

[[16 29]
 [19 28]
 [ 7 34]
 [ 8 27]
 [16 23]]


In [13]:
arr2d1 = np.random.randint(1, 34, (3,3))
print(arr2d1)

[[20 33 32]
 [24  3 14]
 [32 12 14]]


In [14]:
arr2d2 = np.random.randint(31, 56, (3, 3))
print(arr2d2)

[[53 40 46]
 [55 53 41]
 [34 35 45]]


In [15]:
stack2d1 = np.stack([arr2d1, arr2d2])
print(stack2d1)
print(stack2d1.shape)

[[[20 33 32]
  [24  3 14]
  [32 12 14]]

 [[53 40 46]
  [55 53 41]
  [34 35 45]]]
(2, 3, 3)


In [16]:
stack2d2 = np.stack([arr2d1, arr2d2], axis = 1)
print(stack2d2)
print(stack2d2.shape)

[[[20 33 32]
  [53 40 46]]

 [[24  3 14]
  [55 53 41]]

 [[32 12 14]
  [34 35 45]]]
(3, 2, 3)


In [17]:
stack2d3 = np.stack([arr2d1, arr2d2], axis = 2)
print(stack2d3)
print(stack2d3.shape)

[[[20 53]
  [33 40]
  [32 46]]

 [[24 55]
  [ 3 53]
  [14 41]]

 [[32 34]
  [12 35]
  [14 45]]]
(3, 3, 2)


In [18]:
arr2d1 = np.random.randint(1, 34, (2,2))
print(arr2d1)

[[11 24]
 [25  6]]


In [19]:
arr2d2 = np.random.randint(31, 56, (2, 2))
print(arr2d2)

[[55 51]
 [37 47]]


In [20]:
stack2d1 = np.stack([arr2d1, arr2d2])
print(stack2d1)
print(stack2d1.shape)

[[[11 24]
  [25  6]]

 [[55 51]
  [37 47]]]
(2, 2, 2)


In [21]:
stack2d2 = np.stack([arr2d1, arr2d2], axis = 1)
print(stack2d2)
print(stack2d2.shape)

[[[11 24]
  [55 51]]

 [[25  6]
  [37 47]]]
(2, 2, 2)


In [22]:
stack2d2 = np.stack([arr2d1, arr2d2], axis = 2)
print(stack2d2)
print(stack2d2.shape)

[[[11 55]
  [24 51]]

 [[25 37]
  [ 6 47]]]
(2, 2, 2)


## Types of Stacking
<li>There are three types of stacking. They are as follows:</li>
<ol>
    <li><b>Stacking along rows:</b>NumPy provides a helper function: hstack() to stack along rows.</li>
    <li><b>Stcaking along columns:</b>To stack two arrays along columns, Numpy also provides a helper function: vstack().</li>
    <li><b>Stacking along depth:</b>Another helper function to stack two numpy arrays along their depth, we have dstack().</li>
</ol>

In [23]:
arr2d1 = np.random.randint(1,21,(2,2))
print(arr2d1)

[[ 4  5]
 [ 1 11]]


In [24]:
arr2d2 = np.random.randint(22,41,(2,2))
print(arr2d2)

[[39 38]
 [25 33]]


In [29]:
hstack2d = np.hstack([arr2d1, arr2d2])
print(hstack2d)
print(hstack2d.shape)

[[ 4  5 39 38]
 [ 1 11 25 33]]
(2, 4)


In [27]:
np.concatenate([arr2d1, arr2d2], axis = 1)

array([[ 4,  5, 39, 38],
       [ 1, 11, 25, 33]])

In [30]:
vstack2d = np.vstack([arr2d1, arr2d2])
print(vstack2d)

[[ 4  5]
 [ 1 11]
 [39 38]
 [25 33]]


In [34]:
print(np.concatenate([arr2d1, arr2d2], axis= 0))

[[ 4  5]
 [ 1 11]
 [39 38]
 [25 33]]


In [35]:
dstack2d = np.dstack([arr2d1, arr2d2])
print(dstack2d)

[[[ 4 39]
  [ 5 38]]

 [[ 1 25]
  [11 33]]]


### 3. numpy.block():

<li>numpy.block() function is used to create nd-arrays from nested blocks of lists.</li>
<!-- <li>Blocks in the innermost lists are concatenated along the last dimension (-1) first.</li> -->
<!-- <li>Then these blocks gets concatenated along the second-last dimension (-2), and so on until the outermost list is reached.</li> -->

**Syntax**
<code>
    numpy.block(arrays)
</code>

<li>np.block() will raise a value error if:</li>
<ol>
    <li>If list depths are mismatched - for instance, [[a, b], c] is illegal, and should be spelt [[a, b], [c]]</li>
</ol>

In [40]:
onedarr1 = np.array([1,2,3,4,5])
onedarr2 = np.array([6,7,8,9])

In [41]:
block1 = np.block([onedarr1, onedarr2])
print(block1)

[1 2 3 4 5 6 7 8 9]


In [42]:
twodarray1 = np.random.randint(1, 21, (2,2))
twodarray2 = np.random.randint(21, 41, (2,2))
print(twodarray1)
print()
print(twodarray2)

[[ 7 13]
 [17  8]]

[[22 25]
 [29 32]]


In [44]:
block2d = np.block([twodarray1, twodarray2])
print(block2d)
print(block2d.shape)

[[ 7 13 22 25]
 [17  8 29 32]]
(2, 4)


## Splitting NumPy Arrays

<li>Splitting is reverse operation of Joining.</li>

<li>Joining merges multiple arrays into one and Splitting breaks one array into multiple.</li>

<li>We use <b>array_split()</b> for splitting arrays, we pass it the array we want to split and the number of splits.</li>

<li>np.array_split() accepts two parameters. They are:</li>
<ol>
    <li>numpy array </li>
    <li>no of splits</li>
</ol>

<code>
    np.array_split(arr, 2)
</code>

<li>This code will divide a numpy array into 2 splits.</li>
<li>If the array has less elements than required, it will adjust from the end accordingly.</li>
<li>We also have the method split() available but it will not adjust the elements when elements are less in source array for splitting like array_split()</li>

<li>We can use <b>np.hsplit(arr2d, 3)</b> method to split the 2-D array into three 2-D arrays along rows.</li>
<li>We can use <b>np.vsplit(arr2d, 3)</b> method to split the 2-D array into three 2-D arrays along columns.</li>
<li>We can use <b>np.dsplit(arr3d, 3)</b> method to split the 3-D array into three 3-D arrays along depth.</li>




In [45]:
onedarray = np.random.randint(1, 21, 9)
print(onedarray)

[19 12 10 10  1  7 10  7 19]


In [47]:
np.array_split(onedarray, 3)

[array([19, 12, 10]), array([10,  1,  7]), array([10,  7, 19])]

In [48]:
twodarray = np.random.randint(21, 52, (3,3))
print(twodarray)
print(twodarray.shape)

[[50 30 33]
 [46 22 27]
 [28 27 37]]
(3, 3)


In [49]:
np.array_split(twodarray, 3)

[array([[50, 30, 33]]), array([[46, 22, 27]]), array([[28, 27, 37]])]

In [55]:
np.vsplit(twodarray, 3)

[array([[50, 30, 33]]), array([[46, 22, 27]]), array([[28, 27, 37]])]

In [56]:
np.hsplit(twodarray, 3)

[array([[50],
        [46],
        [28]]),
 array([[30],
        [22],
        [27]]),
 array([[33],
        [27],
        [37]])]

In [58]:
threedarray = np.random.randint(1,31,(2,2,2))
print(threedarray.shape)

(2, 2, 2)


In [59]:
print(threedarray)

[[[22 13]
  [12 24]]

 [[18  1]
  [29 13]]]


In [63]:
np.hsplit(threedarray, 2)[0].shape

(2, 1, 2)

In [62]:
np.vsplit(threedarray, 2)[0].shape

(1, 2, 2)

In [64]:
np.dsplit(threedarray, 2)

[array([[[22],
         [12]],
 
        [[18],
         [29]]]),
 array([[[13],
         [24]],
 
        [[ 1],
         [13]]])]

## NumPy Array Functions 
<ol>
    <li><b>Searching</b></li>
    <li><b>Sorting</b></li>
    <li><b>Filtering</b></li>
</ol>

### Search Array:
<li>Numpy provides various methods for searching different kinds of numerical values. They are:</li>
<ol>
    <b><li>numpy.where()</li></b>
    <b><li>numpy.searchsorted()</li></b>
</ol>

#### 1. np.where()
<li>To search an array, use the np.where() method.</li>
<li>You can search an array for a certain value, and return the indexes that get a match.</li>

**Syntax:**

<code>
    numpy.where(condition[, x, y])
</code>

**Parameters:**
<ol>
    <li>condition : When True, yield x, otherwise yield y.</li>
    <li>x, y : Values from which to choose.</li>
</ol>

**Returns:**
<li>[ndarray or tuple of ndarrays] If both x and y are specified, the output array contains elements of x where condition is True, and elements from y elsewhere.</li>






In [69]:
arr1d = np.random.randint(1, 21, 6)
print(arr1d)

[ 2  6 10 17 11 10]


In [66]:
np.where(arr1d < 12)

(array([2, 5], dtype=int64),)

In [67]:
np.where(arr1d < 12, 0, 1)

array([1, 1, 0, 1, 1, 0])

In [72]:
np.where(arr1d == 2)

(array([0], dtype=int64),)

#### Question
<ol>
    <li>Find the indexes where the values are even from the given numpy array.</li>
    <li>Find the indexes where the values are odd from the given numpy array.</li>
</ol>

arr1 = np.array([1,2,3,4,5,6,7,8,9,10])

In [74]:
arr1 = np.array([1,2,3,4,5,6,7,8,9,10])
print(arr1)

[ 1  2  3  4  5  6  7  8  9 10]


In [77]:
np.where(arr1 % 2 == 0, "even", "odd")

array(['odd', 'even', 'odd', 'even', 'odd', 'even', 'odd', 'even', 'odd',
       'even'], dtype='<U4')

In [76]:
np.where(arr1 % 2 == 1)

(array([0, 2, 4, 6, 8], dtype=int64),)

#### 2. np.searchsorted()

<li>The function performs a binary search in an array and returns the index where the specified value would be inserted to maintain the search order.</li>
<li>It is different from np.where() in a way it returns an index such that it maintain the sorted order.</li>

**Syntax :**
<code>
    numpy.searchsorted(arr, num, side=’left’, sorter=None)
</code>

**Parameters :**
numpy.searchsorted(a, v, side='left', sorter=None)



<li>a : [array_like] Input array. If sorter is None, then it must be sorted in ascending order, otherwise sorter must be an array of indices that sort it.</li>

<li>v : [array_like]The Values which we want to insert into arr.</li>
<li>side : [‘left’, ‘right’], optional.</li>
<ol>
<li>If ‘left’, the index of the first suitable location found is given.</li>
<li>If ‘right’, return the last such index.</li>
<li>If there is no suitable index, return either 0 or N (where N is the length of a).</li>
</ol>
<li>sorter: Optional array of integer indices that sort array a into ascending order. They are typically the result of argsort.
np.argsort(arr)</li>

**Return :**
<li>indices, Array of insertion points with the same shape as num.</li>



In [78]:
arr1d = np.array([9,2,5,2,7,3,5,6,2,6])
print(arr1d)

[9 2 5 2 7 3 5 6 2 6]


In [80]:
np.searchsorted(arr1d, 2, side='left', sorter=np.argsort(arr1d))

0

In [81]:
np.searchsorted(arr1d, 2, side='right', sorter=np.argsort(arr1d))

3

In [82]:
np.searchsorted(arr1d, 7, side='right', sorter=np.argsort(arr1d))

9

In [83]:
np.searchsorted(arr1d, 7, side='left', sorter=np.argsort(arr1d))

8

In [87]:
# np.searchsorted(arr1d, 2, side='right')

2

## Sorting NumPy Arrays

<li>Sorting means arranging elements in an ordered sequence either in increasing order or a decreasing order.</li>
<li>Ordered sequence is any sequence that has an order corresponding to elements, like numeric or alphabetical, ascending or descending.</li>
<li>The NumPy ndarray object has a function called <b>sort()</b>, that will sort a specified array.</li>
<li>We can also sort the boolean values using the <b>np.sort()</b> function.</li>
<li>We can also get the indexes of the sorted array using the function <b>np.argsort()</b>.</li>

In [2]:
import numpy as np
arr = np.array([4,2,7,3,9,1,6])
print(arr)

[4 2 7 3 9 1 6]


In [3]:
sorted_arr = np.sort(arr)
print(sorted_arr)

[1 2 3 4 6 7 9]


In [4]:
index_sorted_arr = np.argsort(arr)
print(index_sorted_arr)

[5 1 3 0 6 2 4]


#### Question
array = [9,3,1,6,2,8,4,5]
**<li>Sort the following array in a descending order.</li>**

In [5]:
array = np.array([9, 3, 1, 6, 2, 8, 4, 5])
print(array)

[9 3 1 6 2 8 4 5]


In [6]:
asc_array = np.sort(array)
print(asc_array)

[1 2 3 4 5 6 8 9]


In [9]:
desc_array = - np.sort(-array)
print(desc_array)

[9 8 6 5 4 3 2 1]


#### Filtering NumPy Array

<li>Getting some elements out of an existing array and creating a new array out of them is called filtering.</li>
<li>In NumPy, you filter an array using a boolean index list.</li>
<li>A boolean index list is a list of booleans corresponding to indexes in the array.</li>
<li>If the value at an index is True then that element is contained in the filtered array.</li>
<li>If the value at that index is False that element is excluded from the filtered array.</li>
<li>We can use these comparision operators also to filter the numpy array.</li>
<ol>
    <li>Greater than (>) Or numpy.greater()</li>
    <li>Less Than (<) numpy.less().</li>
    <li>Equal(==) or numpy.equal()</li>
    <li>Not Equal(!=) or numpy.not_equal().</li>
    <li>Greater than and equal to(>=).</li>
    <li>Less than Equal to(<=).</li>


In [10]:
array1 = np.array([4, 5, 3, 2])


In [11]:
array1 < 4

array([False, False,  True,  True])

In [12]:
array1[array1 < 4]

array([3, 2])

In [13]:
array2 = np.array(['Ram', 'Shiva', 'Hari', 'Nirmala'])
bool2 = np.array([True, False, True, False])

In [14]:
array2[bool2]

array(['Ram', 'Hari'], dtype='<U7')

In [27]:
array3 = np.random.randint(1, 31, (3,3))
print(array3)

[[ 7 27 18]
 [ 8 13  8]
 [12  3  6]]


In [28]:
array3[array3 > 15]

array([27, 18])

#### Filter Values Using “OR” Condition

In [29]:
array3 > 20

array([[False,  True, False],
       [False, False, False],
       [False, False, False]])

In [30]:
array3 < 10

array([[ True, False, False],
       [ True, False,  True],
       [False,  True,  True]])

In [31]:
array3[(array3 < 10) | (array3 > 20)]

array([ 7, 27,  8,  8,  3,  6])

#### Filter Values Using "AND" condition

In [34]:
array4 = np.array([[7,27,18], [8,13,8], [12,3,6]])
print(array4)

[[ 7 27 18]
 [ 8 13  8]
 [12  3  6]]


In [38]:
(array4 < 10)

array([[ True, False, False],
       [ True, False,  True],
       [False,  True,  True]])

In [39]:
(array4 %2 == 0)

array([[False, False,  True],
       [ True, False,  True],
       [ True, False,  True]])

In [40]:
array4[(array4 < 10) & (array4 %2 == 0)]

array([8, 8, 6])

#### NumPy Functions

<li>Shuffle</li>
<li>Unique</li>
<li>Resize</li>
<li>Flatten</li>
<li>Ravel</li>

<b>1. Shuffle:</b>
<li>We can get the random positioning of different integer values in the numpy array with the help of numpy.random.shuffle() method</li>
<li>All the values in an array will be shuffled randomly by np.random.suffle() method.</li>

**Syntax :**
<code>
    numpy.random.suffle(x)
</code>
<li>It returns the reshuffled numpy array.</li>

In [46]:
array1d = np.array([1,5,8,2,9])
print(array1d)

[1 5 8 2 9]


In [47]:
np.random.shuffle(array1d)

In [48]:
array1d

array([8, 1, 9, 2, 5])

<b>2. Unique:</b>

<li>The unique() function is used to find the unique elements of an array.</li>
<li>It Returns the sorted unique elements of an array.</li>

**Syntax:**
<code>
    np.unique(arr, return_index=False, return_counts=False)

In [49]:
array2 = np.array([4,5,5,6,2,2,7,3,3])

In [50]:
np.unique(array2)

array([2, 3, 4, 5, 6, 7])

In [51]:
np.unique(array2, return_index = True)

(array([2, 3, 4, 5, 6, 7]), array([4, 7, 0, 1, 3, 6], dtype=int64))

In [52]:
np.unique(array2, return_counts = True)

(array([2, 3, 4, 5, 6, 7]), array([2, 2, 1, 2, 1, 1], dtype=int64))

<b>3. Resize:</b>
<li>The resize() function is used to create a new array with the specified shape.</li>
<li>If the new array is larger than the original array, then the new array is filled with repeated copies of a.</li>
<li>np.reshape() does not changes the original array but only returns the changed array</li>
<li>However, the resize() method returns nothing and directly changes the original array.</li>

In [61]:
array4 = np.array([1,2,3,4,5,6])

In [62]:
array5 = np.resize(array4, (6,2))
print(array5)

[[1 2]
 [3 4]
 [5 6]
 [1 2]
 [3 4]
 [5 6]]


In [59]:
print(array4)

[1 2 3 4 5 6]


<b>Flatten Vs Ravel:</b>
<li>The flatten() function is used to get a copy of an given array collapsed into one dimension.</li>
<li>We use arr.flatten() to collapse ndarray to one-d array.</li>
<li>The ravel() function is used to get a view of the original array collapsed into one dimension.</li>
<li>We use np.ravel() to collapse ndarray to one-d array.</li>
<li>The ravel method returns merely a reference/view of the original array, whereas flattening returns an array's copy.</li>
<li>Ravel requires less time than flatten() to squish NumPy arrays down to 1 Dimension.</li>
<li>The ravel method may be applied to any object that can be effectively parsed, whereas flatten is only applicable to ndarray objects.</li>


In [64]:
array2d = np.random.randint(12, 45, (3,3))
print(array2d)
print(array2d.shape)

[[18 25 21]
 [35 28 24]
 [39 21 18]]
(3, 3)


In [67]:
array2d.flatten()

array([18, 25, 21, 35, 28, 24, 39, 21, 18])

In [69]:
np.ravel(array2d)

array([18, 25, 21, 35, 28, 24, 39, 21, 18])

In [70]:
list_of_list = [[1,2,3], [4,5,6], [7,8,9]]

In [72]:
np.ravel(list_of_list)

array([1, 2, 3, 4, 5, 6, 7, 8, 9])

In [73]:
list_of_list.flatten()

AttributeError: 'list' object has no attribute 'flatten'

#### Insert
<li>We can insert/add elements in a numpy array.</li>
<li>This can be achieved by using a function np.insert()</li>
<li>np.insert() accepts three parameters.</li>
<li>They are as follows:</li>
<ol>
    <li>Numpy arrays</li>
    <li>index position</li>
    <li>values to be inserted</li>
</li>

In [97]:
onedarray = np.array([1,3,5,7,9])
print(onedarray)

[1 3 5 7 9]


In [98]:
onedarray = np.insert(onedarray, 1, 2)

In [99]:
onedarray

array([1, 2, 3, 5, 7, 9])

In [100]:
finalonedarray = np.insert(onedarray, [3,4,5], [4,6,8])

In [101]:
finalonedarray

array([1, 2, 3, 4, 5, 6, 7, 8, 9])

In [ ]:
[1,10,2,3,11,4,5,6,12,7,8,13,9]

In [102]:
np.insert(finalonedarray, [1,3,6,8], [10,11,12,13])

array([ 1, 10,  2,  3, 11,  4,  5,  6, 12,  7,  8, 13,  9])

#### Matrix In A Numpy Array
<li>Matrix is an array of numbers arranged in rows and columns.</li>
<li>We can create a matrix by using np.matrix() method.</li>

**How to create a matrix in numpy?**

<li>For creating a matrix in a numpy array, we can use the following code:</li>
<code>
    matrix1 = np.matrix(numpy_array1)
</code>

In [104]:
matrix1 = np.matrix([[1,2,3,4,5], [6,7,8,9,10]])

In [108]:
print(matrix1)
print(matrix1.ndim)
print(matrix1.shape)
print(type(matrix1))

[[ 1  2  3  4  5]
 [ 6  7  8  9 10]]
2
(2, 5)
<class 'numpy.matrix'>


#### Arithmetic Operations In Matrix

<b>1. Addition</b>
<li>We can add two matrices by using '+' operator.</li>
<li>If m1 and m2 are two numpy matrices then we can calculate the sum of these two matrices using m1 + m2.</li>

In [109]:
m1 = np.matrix([[1,2], [3,4]])
print(m1)
print(m1.shape)

m2 = np.matrix([[5,6], [7,8]])
print()
print(m2)
print(m2.shape)

[[1 2]
 [3 4]]
(2, 2)

[[5 6]
 [7 8]]
(2, 2)


In [111]:
addition_matrix = m1 + m2
print(addition_matrix)
print(addition_matrix.shape)

[[ 6  8]
 [10 12]]
(2, 2)


<b>2. Subtraction</b>
<li>We can subtract two matrices by using '-' operator.</li>
<li>If m1 and m2 are two numpy matrices then we can calculate the difference of these two matrices using m1 - m2.</li>

In [112]:
m1 = np.matrix([[1,2], [3,4]])
print(m1)
print(m1.shape)

m2 = np.matrix([[5,6], [7,8]])
print()
print(m2)
print(m2.shape)

[[1 2]
 [3 4]]
(2, 2)

[[5 6]
 [7 8]]
(2, 2)


In [113]:
m1 - m2 

matrix([[-4, -4],
        [-4, -4]])

<b>3. Matrix Multiplication</b>
<li>We can multiply two matrices by using '*' operator.</li>
<li>If m1 and m2 are two numpy matrices then we can perform matrix multiplication by using m1 * m2.</li>

<li>There are three main ways to perform NumPy matrix multiplication:</li>
<ol>
    <li>np.dot(array a, array b): returns the scalar or dot product of two arrays.</li>
    <li>np.matmul(array a, array b): returns the matrix product of two arrays.</li>
</ol>

In [114]:
m1 = np.matrix([[1,2], [3,4]])
print(m1)
print(m1.shape)

m2 = np.matrix([[5,6], [7,8]])
print()
print(m2)
print(m2.shape)

[[1 2]
 [3 4]]
(2, 2)

[[5 6]
 [7 8]]
(2, 2)


In [115]:
m1 * m2

matrix([[19, 22],
        [43, 50]])

In [116]:
np.matmul(m1, m2)

matrix([[19, 22],
        [43, 50]])

In [117]:
np.dot(m1, m2)

matrix([[19, 22],
        [43, 50]])

In [118]:
np.multiply(m1, m2)

matrix([[ 5, 12],
        [21, 32]])

#### Matrix Functions In NumPy
<ol>
    <b><li>Transpose</li></b>
    <b><li>Swapaxes</li></b>
    <b><li>Inverse</li></b>
    <b><li>Power</li></b>
    <b><li>Determinate</li></b>
</ol>

<b>1. Transpose:</b>
<li>The transpose of a matrix arranges rows into columns and columns into rows.</li>
<li>If the shape of a matrix is 2*3 before transpose then the shape after transpose will be 3 * 2. Rows are converted into columns and columns into rows.</li>

<li>We can transpose a given matrix by using np.transpose() method.</li>

In [119]:
matrix1 = np.matrix([[2, 3, 4], [5,6,7]])
print(matrix1)
print(matrix1.shape)

[[2 3 4]
 [5 6 7]]
(2, 3)


In [120]:
transpose_matrix = np.transpose(matrix1)
print(transpose_matrix)
print(transpose_matrix.shape)

[[2 5]
 [3 6]
 [4 7]]
(3, 2)


<b>2. Swapaxes:</b>
<li>The swapaxes() function is used to interchange two axes of an array.</li>
<li>We can use np.swapaxes() method to swap the axes of an array.</li>

**Syntax**
<code>
    np.swapaxes(matrix, source, destination)
</code>

In [122]:
print(matrix1)
print(matrix1.shape)

[[2 3 4]
 [5 6 7]]
(2, 3)


In [123]:
swaped_matrix = np.swapaxes(matrix1, 0, 1)
print(swaped_matrix)
print(swaped_matrix.shape)

[[2 5]
 [3 6]
 [4 7]]
(3, 2)


<b>3. Matrix Inverse:</b>
<li>The inverse of matrix is another matrix, which on multiplying with the given matrix gives the multiplicative identity.</li>
<li>Inverse of Matrix for a matrix A is denoted by A-1.</li>
<li>The inverse of a 2 × 2 matrix can be calculated using a simple formula.</li>
<li>Further, to find the inverse of a matrix of order 3 or higher, we need to know about the determinant and adjoint of the matrix.</li>
<li>Matrix Inverse (A-1) = 1/det(A) * Adj(A)</li>
<li>We can only calculate the inverse of a square matrix.</li>
<li>In numpy, we can calculate the inverse of a matrix by using np.linalg.inv() method.</li>


In [124]:
matrix1 = np.matrix([[1,2], [7,8]])
print(matrix1)
print(matrix1.shape)

[[1 2]
 [7 8]]
(2, 2)


In [125]:
inverse_matrix1 = np.linalg.inv(matrix1)
print(inverse_matrix1)
print(inverse_matrix1.shape)

[[-1.33333333  0.33333333]
 [ 1.16666667 -0.16666667]]
(2, 2)


<b>4. Power in Matrix</b>

<li>We can calculate the power of a matrix by using np.linalg.matrix_power() method.</li>
<li>It accepts two parameters :</li>
<ol>
    <li>matrix</li>
    <li>n ->power</li>
</ol>
<li>If value of n equals 0 then the matrix power 0 always equal identity matrix.</li>
<li>If value of n > 0 then the matrix multiplication is performed for n times.</li>
<li>If value of n < 0 then the matrix multiplication is done for n times with its inverse.</li>

In [130]:
matrix_order3 = np.matrix([[1,2,3], [4,5,6], [7,8,11]])
print(matrix_order3)

[[ 1  2  3]
 [ 4  5  6]
 [ 7  8 11]]


In [131]:
np.linalg.matrix_power(matrix_order3, 0)

matrix([[1, 0, 0],
        [0, 1, 0],
        [0, 0, 1]])

In [132]:
np.linalg.matrix_power(matrix_order3, 2)

matrix([[ 30,  36,  48],
        [ 66,  81, 108],
        [116, 142, 190]])

In [133]:
np.linalg.matrix_power(matrix_order3, -2)

matrix([[ 1.50000000e+00, -6.66666667e-01, -5.55111512e-17],
        [-3.33333333e-01,  3.66666667e+00, -2.00000000e+00],
        [-6.66666667e-01, -2.33333333e+00,  1.50000000e+00]])

<b>5. Determinant</b>
<li>For a 2×2 matrix (2 rows and 2 columns):</li>

**A = [[a b] [c d]]**

<li>The determinant is:</li>
<code>
    |A| = ad − bc
</code>

**"The determinant of A equals a times d minus b times c"**

<li>In numpy, we can calculate the determinant of a matrix by using np.linalg.det() method.</li>

In [134]:
np.linalg.det(matrix_order3)

-6.0

In [135]:
matrix_order3s = np.matrix([[1,2,3], [4,5,6], [7,8,9]])
print(matrix_order3s)

[[1 2 3]
 [4 5 6]
 [7 8 9]]


In [136]:
np.linalg.det(matrix_order3s)

0.0